PR: https://github.com/odoo/odoo/pull/220378

From: cdc7a76aef0303d4582e15968328e709d2ffdbd0
From: Mazen
Date: 2025-10-07 12:47:18

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = ['product.template.attribute.exclusion', 'pos.load.mixin']

Total Changes: 26

[IMP] product, sale, website_sale, point_of_sale: simplify variant setup

This commit will overhaul product variant setup:
* Allow managing product variants from the list directly
  instead of from a form
* Simplify exclusions: users can now only exclude values on
  the same product and not on optional products
* Allow users to show/hide the extra added price separately
  on the shop page for specific product variants

Users used to be able to exclude values for optional products not just
the same product they're editing, and this was removed to simplify the
flow for setting up variants and also because users could still add
the excluded values of the order product simply by adding them in a
new order line.

Because we don't need to keep track of exclusions across different
products, exclusions were simplified and they are now simply just a
many2many field instead of a whole separate model.

task-4731792

closes odoo/odoo#220378

Related: odoo/enterprise#90929
Related: odoo/upgrade#8116
Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/pos_session.py
+++ b/addons/point_of_sale/models/pos_session.py
@@ -139,7 +139,7 @@ class PosSession(models.Model):
     def _load_pos_data_models(self, config):
         return ['pos.config', 'pos.preset', 'resource.calendar.attendance', 'pos.order', 'pos.order.line', 'pos.pack.operation.lot', 'pos.payment', 'pos.payment.method', 'pos.printer',
             'pos.category', 'pos.bill', 'res.company', 'account.tax', 'account.tax.group', 'product.template', 'product.product', 'product.attribute', 'product.attribute.custom.value',
-            'product.template.attribute.line', 'product.template.attribute.value', 'product.template.attribute.exclusion', 'product.combo', 'product.combo.item', 'res.users', 'res.partner', 'product.uom',
+            'product.template.attribute.line', 'product.template.attribute.value', 'product.combo', 'product.combo.item', 'res.users', 'res.partner', 'product.uom',
             'decimal.precision', 'uom.uom', 'res.country', 'res.country.state', 'res.lang', 'product.category', 'product.pricelist', 'product.pricelist.item',
             'account.cash.rounding', 'account.fiscal.position', 'stock.picking.type', 'res.currency', 'pos.note', 'product.tag', 'ir.module.module', 'account.move']
 

--- a/addons/point_of_sale/models/product_attribute.py
+++ b/addons/point_of_sale/models/product_attribute.py
@@ -55,18 +55,4 @@ class ProductTemplateAttributeValue(models.Model):
 
     @api.model
     def _load_pos_data_fields(self, config):
-        return ['attribute_id', 'attribute_line_id', 'product_attribute_value_id', 'price_extra', 'name', 'is_custom', 'html_color', 'image', 'exclude_for']
-
-
-class ProductTemplateAttributeExclusion(models.Model):
-    _name = 'product.template.attribute.exclusion'
-    _inherit = ['product.template.attribute.exclusion', 'pos.load.mixin']
-
-    @api.model
-    def _load_pos_data_domain(self, data, config):
-        loaded_product_tmpl_ids = list({p['id'] for p in data['product.template']})
-        return [('product_tmpl_id', 'in', loaded_product_tmpl_ids)]
-
-    @api.model
-    def _load_pos_data_fields(self, config):
-        return ['value_ids']
+        return ['attribute_id', 'attribute_line_id', 'product_attribute_value_id', 'price_extra', 'name', 'is_custom', 'html_color', 'image', 'excluded_value_ids']

--- a/addons/point_of_sale/models/product_template.py
+++ b/addons/point_of_sale/models/product_template.py
@@ -114,13 +114,6 @@ class ProductTemplate(models.Model):
         product_tmpl_attr_value = product_tmpls.attribute_line_ids.product_template_value_ids
         product_tmpl_attr_value_read = product_tmpl_attr_value._load_pos_data_read(product_tmpl_attr_value, config)
 
-        # product.template.attribute.exclusion loading
-        product_tmpl_excl = self.env['product.template.attribute.exclusion']
-        product_tmpl_exclusion = product_tmpl_attr_value.exclude_for + product_tmpl_excl.search([
-            ('product_tmpl_id', 'in', product_tmpls.ids),
-        ])
-        product_tmpl_exclusion_read = product_tmpl_excl._load_pos_data_read(product_tmpl_exclusion, config)
-
         # product.product loading
         product_read = products._load_pos_data_read(products.with_context(display_default_code=False), config)
 
@@ -157,7 +150,6 @@ class ProductTemplate(models.Model):
             'product.combo.item': combo_item_read,
             'product.template.attribute.value': product_tmpl_attr_value_read,
             'product.template.attribute.line': product_tmpl_attr_line_read,
-            'product.template.attribute.exclusion': product_tmpl_exclusion_read,
         }
 
     @api.model
